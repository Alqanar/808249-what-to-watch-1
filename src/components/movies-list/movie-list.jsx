import React, {PureComponent} from "react";
import PropTypes from "prop-types";

import FilmCard from "../film-card/film-card.jsx";


class MovieList extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeCardId: null
    };

    this._handleCardHover = this._handleCardHover.bind(this);
  }

  render() {
    const {films, onClick} = this.props;
    const {activeCardId} = this.state;

    return (
      <div className="catalog__movies-list">
        {films.map((film) => <FilmCard
          movie={film}
          key={film.id}
          onClick={onClick}
          onMouseEnter={this._handleCardHover}
          isActive={film.id === activeCardId}
        />
        )}
      </div>
    );
  }

  _handleCardHover(id) {
    this.setState({
      activeCardId: id
    });
  }
}

MovieList.propTypes = {
  films: PropTypes.array.isRequired,
  onClick: PropTypes.func
};

export default MovieList;
