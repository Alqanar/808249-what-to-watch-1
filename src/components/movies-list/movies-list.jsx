import React from "react";
import PropTypes from "prop-types";

import FilmCard from "../film-card/film-card.jsx";


const MoviesList = (props) => {
  const {films, onClick} = props;

  const filmsList = films.map((film) => (
    <FilmCard
      movie={film}
      key={film.id}
      onClick={onClick}
    />
  ));

  return (
    <div className="catalog__movies-list">
      {filmsList}
    </div>
  );
};

MoviesList.propTypes = {
  films: PropTypes.array.isRequired,
  onClick: PropTypes.func
};

export default MoviesList;
