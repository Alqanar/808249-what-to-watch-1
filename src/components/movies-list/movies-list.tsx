import * as React from "react";
import {connect} from "react-redux";

import FilmCard from "../film-card/film-card";
import {getFilterFilms} from "../../reducer/movie/selectors.js";
import {moviesListMock} from "../main-page/test-mock-data.js";
import {IFilm} from "../../types";


interface IProps {
  films: IFilm[];
  onClick: (films: IFilm) => void;
  useAllFilms: boolean;
  fiteredGenre?: string[];
  limit?: number;
  excludeFilmId?: string;
  onMoreButtonClick: () => void;
  currentLength: number;
  resetCurrentLength: () => void;
}

class MoviesList extends React.PureComponent<IProps, null> {
  public constructor(props) {
    super(props);
  }

  public componentDidUpdate(oldProps: IProps): void {
    const {films, resetCurrentLength} = this.props;

    if (films.length !== oldProps.films.length) {
      resetCurrentLength();
    }
  }

  public render(): React.ReactElement {
    const {onMoreButtonClick, films, currentLength = films.length} = this.props;

    return (
      <>
        <div className="catalog__movies-list">
          {this.filmsList}
        </div>
        <div className={`catalog__more ${currentLength >= films.length ? `visually-hidden` : ``}`}>
          <button onClick={onMoreButtonClick} className="catalog__button" type="button">Show more</button>
        </div>
      </>
    );
  }

  private get filmsList(): React.ReactElement {
    const {
      films,
      onClick,
      useAllFilms,
      fiteredGenre,
      limit,
      excludeFilmId,
      currentLength = films.length
    } = this.props;

    const list = useAllFilms ? films : moviesListMock;
    const filmsLimit = limit ? limit : currentLength;

    return list
      .filter(({id, genre}): boolean => {
        if (id === excludeFilmId) {
          return false;
        }
        if (!fiteredGenre) {
          return true;
        }
        return genre
          .some((item): boolean => fiteredGenre
            .some((filtereGenreItem): boolean => item === filtereGenreItem));
      })
      .splice(0, filmsLimit)
      .map((film): React.ReactElement => (
        <FilmCard
          film={film}
          key={film.id}
          onClick={onClick}
        />
      ));
  }
}


const mapStateToProps = (state, ownProps): void => ({
  ...ownProps,
  films: getFilterFilms(state)
});

export {MoviesList};

export default connect(mapStateToProps)(MoviesList);
