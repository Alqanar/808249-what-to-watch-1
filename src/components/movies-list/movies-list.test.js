import React from "react";
import renderer from "react-test-renderer";

import {MoviesList} from "./movies-list.tsx";

const films = [
  {
    id: `0`,
    name: `The Aftermath`,
    posterLink: `https://st.kp.yandex.net/images/film_iphone/iphone360_1000125.jpg`,
    trailer: `https://youtu.be/FPv3e2FZOgo`
  },
  {
    id: `1`,
    name: `The Professor and the Madman`,
    posterLink: `https://st.kp.yandex.net/images/film_iphone/iphone360_996027.jpg`,
    trailer: `https://youtu.be/ESYU-bkmxuI`
  },
  {
    id: `2`,
    name: `Five Feet Apart`,
    posterLink: `https://st.kp.yandex.net/images/film_iphone/iphone360_1151373.jpg`,
    trailer: `https://youtu.be/XtgCqMZofqM`
  }
];

it(`Movie list correctly renders`, () => {
  const tree = renderer
    .create(<MoviesList
      films={films}
      useAllFilms={true}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
