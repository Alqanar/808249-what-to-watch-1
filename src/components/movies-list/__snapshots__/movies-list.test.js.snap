// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie list correctly renders 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        loop={true}
        muted={true}
        poster="https://st.kp.yandex.net/images/film_iphone/iphone360_1000125.jpg"
        src="https://youtu.be/FPv3e2FZOgo"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Aftermath
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        loop={true}
        muted={true}
        poster="https://st.kp.yandex.net/images/film_iphone/iphone360_996027.jpg"
        src="https://youtu.be/ESYU-bkmxuI"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Professor and the Madman
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        loop={true}
        muted={true}
        poster="https://st.kp.yandex.net/images/film_iphone/iphone360_1151373.jpg"
        src="https://youtu.be/XtgCqMZofqM"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Five Feet Apart
      </a>
    </h3>
  </article>
</div>
`;
